user_app.go

```diff
@@ -72,21 +72,18 @@ func (ledger *LedgerCosmos) Close() error {
 	return ledger.api.Close()
 }
 
-// VersionIsSupported returns true if the App version is supported by this library
+// CheckVersion returns true if the App version is supported by this library
 func (ledger *LedgerCosmos) CheckVersion(ver VersionInfo) error {
-	version, err := ledger.GetVersion()
-	if err != nil {
-		return err
-	}
 
-	switch version.Major {
-	case 1:
-		return CheckVersion(ver, VersionInfo{0, 1, 5, 1})
-	case 2:
-		return CheckVersion(ver, VersionInfo{0, 2, 1, 0})
-	default:
-		return fmt.Errorf("App version is not supported")
+	for _, minSupportedVersion := range minSupportedVersions {
+		if ver.AppMode == minSupportedVersion.AppMode && ver.Major == minSupportedVersion.Major {
```

Comment:
Using appmode to separate app implementations is a good idea but I am not sure about the current checkversion implementation.

The problem is that this PR will result in gaiacli  / CosmosHub CLI to accept Terra app as compatible. Or Terra to accept a Cosmos app.

IMO, an adequate generic checkVersion should accept something like `minSupportedVersions = []VersionInfo`. 

The other problem is that currently Cosmos App requires different behavior for 1.x.x and 2.x.x and probably something different for Terra too. I am tempted to say that in the long term, Terra might be better off using a fork that complies with the same interface but adjust to Terra specific versions/behaviours.

Reply:
I will block it to avoid gaia from accepting the Terra app and viceversa.

Reply:
If we can get appmode as an input from cosmos-sdk's config at calling FindLedgerCosmosUserApp, we can avoid this situation.